### PR TITLE
Fix #4772:Role based provisioning not works as expected

### PR DESCRIPTION
**Pull request by IsurangaPerera**
### Proposed changes in this pull request
- Fix wso2/product-is#4772

**Scenario: Role-based provisioning is enabled and only users with role_x are permitted to be provisioned**

When a user is created with role_x assigned then the preAddUser event is triggered and the provisioning process initiates. At this stage, existing implementation tries to fetch all user roles from the user store manager. As user roles are not yet persisted only **internal/everyone** role is applied to the user being provisioned. Hence role-based provisioning not works for the user even though he has intended roles assigned to him.

In this fix we consider roles from the provisioning entity as well as roles fetched from the user store.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labelled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Comment by emswbandara** [comment]
@IsurangaPerera could you please update the description on the problem observed and the approach taken to resolve the case

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 1]
```diff
@@ -825,7 +827,10 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
         }
 
         String userName = getUserName(provisioningEntity.getAttributes());
-        List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
+        List<String> roleListOfUser = Stream.concat(getUserRoles(userName, tenantDomain).stream(),
+                provisioningEntity.getAttributes().get(ClaimMapping.build
```
Comment (ruwanta) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: what if 
provisioningEntity.getAttributes() is null ?
provisioningEntity.getAttributes().get(x) is null ?

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 2]
```diff
@@ -824,12 +826,22 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
...
+            String userName = getUserName(provisioningEntity.getAttributes());
+            List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
+
+            List<String> newRoleListOfUser = provisioningEntity.getAttributes().get(ClaimMapping.build
+                    (IdentityProvisioningConstants.GROUP_CLAIM_URI, null, null, false));
 
-        for (String provisionByRole : provisionByRoleList) {
-            if (roleListOfUser.contains(provisionByRole)) {
-                return true;
+            if(newRoleListOfUser != null) {
```
Comment (malithie) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: Apply formatting

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 3]
```diff
@@ -824,12 +826,22 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
             return true;
         }
 
-        String userName = getUserName(provisioningEntity.getAttributes());
-        List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
+        if(provisioningEntity.getAttributes() != null) {
```
Comment (emswbandara) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: formatting issue

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 4]
```diff
@@ -824,12 +826,22 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
...
-        for (String provisionByRole : provisionByRoleList) {
-            if (roleListOfUser.contains(provisionByRole)) {
-                return true;
+            if(newRoleListOfUser != null) {
+                 roleListOfUser = Stream.concat(roleListOfUser.stream(), newRoleListOfUser.stream()).distinct()
+                         .collect(Collectors.toList());
+            }
+
+            for (String provisionByRole : provisionByRoleList) {
+                if (roleListOfUser.contains(provisionByRole)) {
```
Comment (emswbandara) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: shall we use ListUtils here?

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 5]
```diff
@@ -824,12 +826,22 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
...
+                    return true;
```
Comment (emswbandara) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: shall we add a debug log here?

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java (hunk 4 above)
Comment (ruwanta) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: We are just checking a particular role is available in two another list?
If so, it is not good to create another list with distinct strings. This creates large list, and add garbage pressure.
Best is to check on original two lists one after another, which does not create new list.

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java [hunk 6]
```diff
@@ -1052,4 +1059,23 @@ private void updateMappedGroupForAttribute(ProvisioningEntity provisioningEntity
...
+     *
+     * @param userRoles
+     * @param provisioningRoles
+     */
+    private boolean userHasProvisioningRoles(List<String> userRoles, List<String> provisioningRoles, String userName) {
+
+        if (CollectionUtils.isNotEmpty(userRoles) && CollectionUtils.isNotEmpty(provisioningRoles) && !ListUtils
+                .intersection(userRoles, Arrays.asList(provisioningRoles)).isEmpty()) {
+            if (log.isDebugEnabled()) {
+                log.debug("User with userName " + userName + " has provisioning role(s) assigned.");
```
Comment (ruwanta) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: General pattern for log is like this(please note colon)
  log.debug("User with userName : " + userName + " has provisioning role(s) assigned.");

Reason, lets assume a user having a name "not found.". Can you imagine what is the log line printed.

Comment (ruwanta) [reply components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java]: Also 
ListUtils.intersection() still creates another which is unnecessary for the purpose

**Pull request merged**